ls1intum/Artemis
# [Feature] Zip Download of programming exercise material (repos, problem statement, details)

### Checklist

### Motivation and Context

This "Download Exercise" feature allows to export the following material of an programming exercise as one zip folder:
- Problem Statement as .md file
- Details (i.e. the programming exercise object) as .json file
- Instructor Repositories: template, solution, tests (as zip)

Closes #2553

### Description
On the **client side**, there was one button added in the exercise detail page (next to download repositories) triggering invoking `exportInstructorExercise` method on `programmingExerciseService` -> sends a GET request to `api/programming-exercise/{exercise-id}/export-instructor-exercise`

On the **server side**, I introduced the respective endpoint and the new service method `ProgrammingExerciseExportService.exportInstructorProgrammingExercise` which downloads the material which is described above.
As it includes the task of downloading repositories and zip file, which was implemented before, I refactored the existing method `exportProgrammingExercise` of the `programmingExerciseExportService`, now called `exportProgrammingExerciseRepositories`: With a new boolean flag `includingStudentRepos` in the method's signature, the callee decides if the students' repositories should be included in the downloaded zip folder (true in the case of exporting the repositories in the `CouseExamExportService`) or only the instructor repositories, i.e. template, solution, tests (in the case of exporting the programming exercise) through new code introduced with this PR. 

### Answered questions
- Should the exercise download only be allowed for instructors? -> yes
- Should the exercise details export include the whole object or should I rather write selected attributed into a file? The lifecycle including the different dates (of type `ZoneTypedDate`) make this file blowing up a lot. -> use `MappingJackson2HttpMessageConverter` to get the object mapper

### Steps for Testing


1. Log in to Artemis as **Instructor**
2. Navigate to Course Overview
3. Select a course
4. Select a programming exercise within in course -> redirect to exercise details page (screencast start)
5. Click the button "Download Exercise"
6. Wait for the success message and the automatic opening of a "where-to-save-the-downlaoded-zip"-diaglog
7. Choose a location and wait for the download to be finished
8. Can it be unzipped? Check the downloaded zip file starting with `Material-`:
8.1: Is there a `Problem-Statement-*.md` file including the statement in markdown format?
8.2 Is there a `Exercise-Details-*-.json` file including the exercise object in valid json?
8.3 Is there a zip folder including the instructor repositories? Can it be unzipped? 
8.3.1 Is there a `-solution` repo? Can it be unzipped?
8.3.2 Is there a `-tests` repo? Can it be unzipped?
8.3.3 Is there a `-exercise` repo? Can it be unzipped?

9. Log in as tutor and do 2.-4.
10. Check if the button "Download Exercise" is not present (as of now, only instructors are allowed).

As download repos as also touched, check if the in step 5 the button "Download Repos" is also possible (student participation is required, as it downloads the students repositories).

### Test Coverage


 
ProgrammingExerciseExportService.java: 72% (new parts covered 100%)
ProgrammingExerciseResource.java: 91%
FileService.java: 82% (catch IOException blocks not covered, increasing coverage for service as a whole should be addressed in a separate PR)
programming-exercise.service.ts 68.42% (new part covered)
programming-exercise-instructor-exercise-download.component.ts: 100%

### Screenshots


Note: the browser dialog appearing in the video in which one could determine where to save the zip file does occur in the brave browser I used, but not in the chrome browser where it is automatically downloaded to a certain folder (e.g. downloads).
![Download-Exercise](https://user-images.githubusercontent.com/41366522/116056586-3c23c780-a67e-11eb-8c6a-5aceeb200824.gif)

The .json file exporting the programming exercise object can be seen here in more detail:
![Folie1](https://user-images.githubusercontent.com/41366522/117427382-461eb380-af25-11eb-9140-7e9969d96792.png)
